### DISCUSSION TITLE
Set parameters regarding antsMultivariateTemplateConstruction2.sh

### DISCUSSION THREAD

**Discussion by zoelir729** [ANTsX/ANTs#1347] [General]
According to some papers, they create template with the following descriptions about the parameters: 

A maximum of 90 iterations at the coarsest level, 30 at the next coarsest, and 90 at full resolution; template update step size was set to 0.1 mm. Four rigid-then-affine iterations were followed by six full runs of the above nonlinear multiresolution routine to ensure stable convergence. (Acosta-Cabronero et al. J Neurosci 2016;36:364–374)
or
For template creation, radio-frequency bias corrected MPRAGE images were spatially normalized using the ‘Greedy-SyN’ approach in ANTs v2.1 with a maximum of 90 x 30 x 30 multi-resolution iterations and template-update step size set to 0.1 mm. Rigid plus-affine initial alignment was followed by six full runs of the above routine. (Acosta-Cabronero et al. Brain 2017;140:118–131)

**And I am not sure the parameters regarding to the sentence "Four rigid-then-affine iterations were followed by six full runs of the above nonlinear multiresolution routine to ensure stable convergence" or "Rigid plus-affine initial alignment was followed by six full runs of the above routine".**
And I read the usage of the antsMultivariateTemplateConstruction2.sh, found the **`-l`** and **`-i`** could explain the sentences. But I am not sure.
`-i`:  Iteration limit (default 4): iterations of the template construction.
`-l`: Use linear image registration stages during the pairwise (template/subject) deformable registration. Otherwise, registration is limited to SyN orB-spline SyN (see '-t' option). This is '1' by default.

Whether the sentence "Four rigid-then-affine iterations...six runs of ..." was written in **`-i 6 -l 4`** and "Rigid plus-affine...six runs of ..." was written in **`-i 6 -l 1`** respectively?

Thanks!
-zoe

**Comment by ntustison** [comment]
I'm assuming the authors are referring to the ``-r`` option.  So my guess would be they mean ``-r 1 -i 6`` and that the ``-l`` option isn't being changed from the default value of ``1``.

**Comment by zoelir729** [comment]
> 



> I'm assuming the authors are referring to the `-r` option. So my guess would be they mean `-r 1 -i 6` and that the `-l` option isn't being changed from the default value of `1`.

How to explain the "**Four** rigid-then-affine iterations"?
It confuses me the most.

**Comment by cookpa** [comment]
I think this means they ran the template script multiple times, using the output of the previous run as input to the next one with the `-z` option. 

I often do this to get a more stable solution. Something like:

  `antsMultivariateTemplateConstruction2.sh -t Affine -i 4 -o myAffine ... `

Then if the output of the above is myAffinetemplate0.nii.gz, in a new directory, run

  `antsMultivariateTemplateConstruction2.sh -z myAffinetemplate0.nii.gz -t SyN -i 6 -m 90x30x30 -o myDeformable ...`

**Reply by zoelir729** [reply]
Thanks for your detailed reply. @cookpa 

So if I have sub001_T1.nii.gz to sub100_T1.nii.gz. (N=100)
To create template accoring to the above description: A maximum of 90 iterations at the coarsest level, 30 at the next coarsest, and 90 at full resolution; template update step size was set to 0.1 mm. Four rigid-then-affine iterations were followed by six full runs of the above nonlinear multiresolution routine to ensure stable convergence. (Acosta-Cabronero et al. J Neurosci 2016;36:364–374)

Step 1 I should write like:
`antsMultivariateTemplateConstruction2.sh -­d 3 -­t Affine -i 4 -o MyAffine sub*.nii.gz`

then according to the usage:
-q:  Max iterations for each pairwise registration (default = 100x100x70x20)
    specified in the form ...xJxKxL where
    - J = max iterations at coarsest resolution (here, reduced by power of 2^2)
    - K = middle resolution iterations (here, reduced by power of 2)
    - L = fine resolution iterations (here, full resolution).
-g:  Gradient step size (default 0.25): smaller in magnitude results in more cautious steps.  Use smaller steps to refine template details. 0.25 is an upper (aggressive) limit for this parameter.

Step 2 like:
`antsMultivariateTemplateConstruction2.sh -z myAffinetemplate0.nii.gz -t SyN -i 6 -g 0.1 -q 90x30x90 -o myDeformable`

Additionally, what the difference between `-q 100x90x30x90` and `-q 90x30x90`? or they are the same.

**Reply by cookpa** [reply]
I don't know how closely you want to follow their procedure, the only way to know exactly what they did is to get their code (and even that [might not be sufficient](https://www.frontiersin.org/articles/10.3389/fninf.2019.00001/full)). They may have used the older `antsMultivariateTemplateConstruction.sh`, which has different options. In fact, now that I think about it, they probably did use the older script, because the older registration code ran faster, and 90 iterations will take a long time at full resolution.

Based on my educated guesses, I would start by preprocessing the images. Usually this means bias correct, remove outlier intensities, and if needed, padding the input images (see usage for ImageMath, PadImage). The padding is needed if the image FOV is tight to the bounding box of the image, you want your images and template to be centered with empty space around them, so that the deformation doesn't run into problems at the edges.

The last thing is to check that the images will make a sensible average, meaning either they have a similar contrast range, or they have a similar range after dividing by the mean (test with ImageMath Normalize).

Then something like:

`antsMultivariateTemplateConstruction2.sh -­d 3 -n 0 -­t Affine -A 0 -i 4 -o MyAffine sub*.nii.gz`

Then check the the template and the warped images, look for any gross failures. The template brain will be positioned at a point around the average of the input physical space - if it's shifted close to the edge, you can re-center or pad more as needed.

Then on to deformable registration,

`antsMultivariateTemplateConstruction2.sh -­d 3 -n 0 -­t SyN -i 6 -o MyAffine sub*.nii.gz`

I'm suggesting to start with the defaults here, but you can adjust

`     -q:  Max iterations for each pairwise registration (default = 100x100x70x20) `

I'm sorry the usage is a bit wrong here (copied from the older script), the coarseness of the levels is controlled by you with -f. I'll update it.

`     -f:  Shrink factors (default = 6x4x2x1):  Also in the same form as -q max iterations.
          Needs to have the same number of components.
          
     -s:  Smoothing factors (default = 3x2x1x0):  Also in the same form as -q max
          iterations.  Needs to have the same number of components.`

In the older script, -f and -s were hidden from the user and -f was fixed at powers of 2. So if you said `-m 100x100x50x20`, it would use shrink factors of 8x4x2x1.

**Reply by zoelir729** [reply]
Thanks, it helps me a lot.

**Comment by cookpa** [comment]
Also "template-update step size set to 0.1 mm" - the update step size is not in mm, it's a dimensionless scaling factor. The default is set to 0.25. The usage advice is to reduce this for more cautious updates, but I'm not sure in practice how much influence it has.

**Comment by cookpa** [comment]
Going to move this to discussions, please feel free to continue there

**Comment by tangYING2022** [comment]
Hi，l have a related question. I also want to create a template with antsMultivariateTemplateConstruction2.sh, but my computer just have one cpu core, the process takes too long time。I don't know if SEG can be installed on a computer with just one cpu core?  or  how can I set parameters to accelerate the speed？
    Additionally,  I have saw an answer  _“You can also run each iteration (the -i option) as a separate run using the template from the previous iteration as your starting template (the -z option). ”_    I'm a little confused what is idea or principle of each iteration？
     After creating the template，should we run antsRegistrationSyNQuick.sh to normalize images to the template?
    Since I am very new to ANTs, I'm looking forward to your reply.
Thanks
--tang

**Reply by ntustison** [reply]
Template building is a computationally intensive process so a single CPU machine might be prohibitive.  Have you thought about using one our [preexisting templates](https://figshare.com/articles/dataset/ANTs_ANTsR_Brain_Templates/915436)?

**Comment by tangYING2022** [comment]
Thanks for your reply，maybe it’s a good idea.     I have another question, what are the differences between antsRegistrationSyNQuick.sh and antsRegistrationSyN.sh?    I find the former is quick, so what causes the difference?-------- 原始邮件 --------发件人： Nick Tustison ***@***.***>日期： 2023年3月20日周一 凌晨4:36收件人： ANTsX/ANTs ***@***.***>抄送： tangYING2022 ***@***.***>, Comment ***@***.***>主    题： Re: [ANTsX/ANTs] Set parameters regarding antsMultivariateTemplateConstruction2.sh (Discussion #1347)
Template building is a computationally intensive process so a single CPU machine might be prohibitive.  Have you thought about using one our preexisting templates?

**Reply by ntustison** [reply]
``Quick`` reduces the number of iterations at the higher resolution levels and uses the mutual information as the similarity metric (vs. cross correlation).

**Comment by tangYING2022** [comment]
I see. Thanks a lot!-------- 原始邮件 --------发件人： Nick Tustison ***@***.***>日期： 2023年3月20日周一 晚上10:11收件人： ANTsX/ANTs ***@***.***>抄送： tangYING2022 ***@***.***>, Comment ***@***.***>主    题： Re: [ANTsX/ANTs] Set parameters regarding antsMultivariateTemplateConstruction2.sh (Discussion #1347)
Quick reduces the number of iterations at the higher resolution levels and uses the mutual information as the similarity metric (vs. cross correlation).
